File: src/SpellUtils.ts
import { Optional } from "KEndec";
import { EnterScopeInstruction, ExitScopeInstruction, SpellInstruction } from "./spellInstruction";
import { Fragment } from "./fragment/fragment";
import { SpellPartData } from "./fragment/trickster/spellPart";
import * as spellPart from "./fragment/trickster/spellPart";

export class SpellUtils {
    static decodeInstructions(instructions: Array<SpellInstruction>, scope: Array<number>, inputs: Array<Fragment<unknown>>, overrideReturnValue: Optional<Fragment>): SpellPart {
        var children: Array<Fragment<SpellPartData>> = [];
        instructions = [...instructions];
        scope = [...scope];

        for (const input of inputs) {
            children.push(spellPart.of(input, []));
        }

        while (instructions.length > 0) {
            var inst = instructions.pop();

            if (inst instanceof EnterScopeInstruction) {
                scope.push(0);
            } else if (inst instanceof ExitScopeInstruction) {
                scope.pop();

                if (scope.length != 0)
                    scope.push(scope.pop()! + 1);
            } else {
                let args;
                {
                    let _args: Array<Fragment<SpellPartData>> = [];
                    for (let i = scope[scope.length - 1]; i > 0; i--)
                        _args.push(children.pop() as Fragment<SpellPartData>);
                    args = _args.toReversed();
                }

                children.push(spellPart.of(inst as Fragment<unknown>, args));
            }
        }

        return overrideReturnValue.map(fragment => spellPart.of(fragment, children.toReversed())).orElse(children.pop())!;
    }

    static flattenNode(head: Fragment<SpellPartData>): Array<SpellInstruction> {
        const instructions: Array<SpellInstruction>  = [];
        const headStack: Array<Fragment<SpellPartData>> = [];
        const indexStack: Array<number> = [];

        headStack.push(head);
        indexStack.push(-1);

        while (headStack.length != 0) {
            let currentNode: Fragment<SpellPartData> = headStack[headStack.length - 1];
            let currentIndex = indexStack.pop()!;

            if (currentIndex == -1) {
                instructions.push(new ExitScopeInstruction());
                instructions.push(currentNode.data.glyph);
            }

            currentIndex++;

            if (currentIndex < currentNode.data.subParts.length) {
                headStack.push(currentNode.data.subParts.toReversed()[currentIndex]);
                indexStack.push(currentIndex);
                indexStack.push(-1);
            } else {
                headStack.pop();
                instructions.push(new EnterScopeInstruction());
            }
        }

        return instructions;
    }
}

File: site/src/jsMain/kotlin/me/maplesyrum/tricksterStudio/SpellUtils.kt
package me.maplesyrum.tricksterStudio

import me.maplesyrum.tricksterStudio.spell.executer.EnterScopeInstruction
import me.maplesyrum.tricksterStudio.spell.executer.ExitScopeInstruction
import me.maplesyrum.tricksterStudio.spell.executer.SpellInstruction
import me.maplesyrum.tricksterStudio.spell.fragment.Fragment
import me.maplesyrum.tricksterStudio.spell.fragment.SpellPart
import tree.maple.kendec.util.Optional

object SpellUtils {
    fun decodeInstructions(
        instructions: List<SpellInstruction>,
        scope: List<Int>,
        inputs: List<Fragment>,
        overrideReturnValue: Optional<Fragment>
    ): SpellPart {
        val children: ArrayDeque<SpellPart> = ArrayDeque<SpellPart>()
        val instructions = ArrayDeque(instructions)
        val scope = ArrayDeque(scope)

        for (input in inputs) {
            children.add(SpellPart(input))
        }

        while (instructions.isNotEmpty()) {
            val inst = instructions.removeLast()

            if (inst is EnterScopeInstruction) {
                scope.add(0)
            } else if (inst is ExitScopeInstruction) {
                scope.removeLast()

                if (!scope.isEmpty()) scope.add(scope.removeLast() + 1)
            } else {
                val args: List<SpellPart>
                run {
                    val _args: ArrayList<SpellPart> = ArrayList<SpellPart>()
                    for (i in scope.last() downTo 1) _args.add(children.removeLast())
                    args = _args.reversed()
                }

                children.add(SpellPart(inst as Fragment, args))
            }
        }

        return overrideReturnValue.map { fragment -> SpellPart(fragment, ArrayList(children).reversed()) }
            .orElse(children.removeLast())!!
    }

    fun flattenNode(head: SpellPart): ArrayDeque<SpellInstruction> {
        val instructions: ArrayDeque<SpellInstruction> = ArrayDeque()
        val headStack: ArrayDeque<SpellPart> = ArrayDeque()
        val indexStack: ArrayDeque<Int> = ArrayDeque()

        headStack.add(head)
        indexStack.add(-1)

        while (!headStack.isEmpty()) {
            val currentNode: SpellPart = headStack.last()
            var currentIndex: Int = indexStack.removeLast()

            if (currentIndex == -1) {
                instructions.add(ExitScopeInstruction())
                instructions.add(currentNode.glyph)
            }

            currentIndex++

            if (currentIndex < currentNode.subParts.size) {
                headStack.add(currentNode.subParts.reversed()[currentIndex])
                indexStack.add(currentIndex)
                indexStack.add(-1)
            } else {
                headStack.removeLast()
                instructions.add(EnterScopeInstruction())
            }
        }

        return instructions
    }
}
